chainerx_cc/chainerx/routines/normalization.cc

```diff
@@ -117,43 +282,61 @@ Array BatchNorm(
+    // Preprocess inputs.
     PreprocessBatchNormResult result = PreprocessBatchNorm(x, gamma, beta, running_mean, running_var, axis);
-    std::shared_ptr<BatchNormForwardBackward> fb =
-            x.device().GetBatchNormForwardBackward(result.mean, result.var, eps, decay, result.sorted_axis);
-
     const Array& gamma_reshaped = result.gamma;
     const Array& beta_reshaped = result.beta;
+    const Array& mean_reshaped = result.mean;
+    const Array& var_reshaped = result.var;
+    const Axes& sorted_axis = result.sorted_axis;
+
+    Device& device = x.device();
+
+    // Compute forward.
+    Array out{};
+    std::shared_ptr<BatchNormGradState> state{};
+    {
+        NoBackpropModeScope scope{};
+        std::tie(out, state) = device.backend().CallOp<BatchNormOp>(
+                x, gamma_reshaped, beta_reshaped, mean_reshaped, var_reshaped, eps, decay, sorted_axis, true, nonstd::nullopt);
+    }
+    CHAINERX_ASSERT(state != nullptr);
 
-    Array out = fb->Forward(x.AsGradStopped(), gamma_reshaped.AsGradStopped(), beta_reshaped.AsGradStopped());
-    CHAINERX_ASSERT(out.dtype() == x.dtype());
     internal::MakeViewForForwardBackwardOutput(out);
 
     BackwardBuilder bb{"batch_norm", {x, gamma_reshaped, beta_reshaped}, {out}};
     if (BackwardBuilder::Target bt = bb.CreateTarget({0, 1, 2})) {
-        bt.Define([fb = std::move(fb), x_tok = bb.RetainInput(0), gamma_tok = bb.RetainInput(1), eps, sorted_axis = result.sorted_axis](
-                          BackwardContext& bctx) {
+        bt.Define([state = std::move(state),
+                   x_tok = bb.RetainInput(0),
+                   gamma_tok = bb.RetainInput(1),
+                   eps,
+                   sorted_axis,
+                   beta_shape = beta_reshaped.shape(),
+                   beta_dtype = beta_reshaped.dtype()](BackwardContext& bctx) {
             const Array& gout = *bctx.output_grad();
+            const Array& x = bctx.GetRetainedInput(x_tok);
+            const Array& gamma_reshaped = bctx.GetRetainedInput(gamma_tok);
+
+            Device& device = x.device();
 
+            // Compute backward.
             Array gx{};
             Array ggamma{};
             Array gbeta{};
             {
-                std::array<Array, 3> ginputs = fb->Backward(gout.AsGradStopped());
-                internal::MakeViewForForwardBackwardOutput(ginputs);
-
-                gx = std::move(ginputs[0]);
-                ggamma = std::move(ginputs[1]);
-                gbeta = std::move(ginputs[2]);
+                NoBackpropModeScope scope{};
+                std::tie(gx, ggamma, gbeta) = device.backend().CallOp<BatchNormGradOp>(
+                        x, gamma_reshaped, gout, eps, sorted_axis, state, nonstd::nullopt, nonstd::nullopt, nonstd::nullopt);
             }
+            internal::MakeViewForForwardBackwardOutput(gx);
+            internal::MakeViewForForwardBackwardOutput(ggamma);
+            internal::MakeViewForForwardBackwardOutput(gbeta);
```

Comment:
Hmm, why are these 3 lines needed?

Reply:
I don't think so actually since no state creation is involved. Let me remove them.

Reply:
Fixed.